### PR TITLE
Build requestdefinition on access, not constructor

### LINKED ISSUE #23: Subclasses of Consumers Not Behaving as Expected
Consider the following scenario:

```python
class StoreAPI(Consumer):
    @get("/api/item?id={id}")
    def get_item(self, id):
         pass

class ProtectedStoreAPI(StoreAPI):
    def __init__(self, user_id):
        super().__init__()
        self.user_id = user_id

    def get_item(self, id):
         if allowed(self.user_id):
             return super().get_item(id)
```

In the above scenario, I attempt to implement a "wrapper" around a consumer. The wrapper, being a subclass, can still be passed around just as the consumer was, but the wrapper has some extra security features.

The problem arises, however, when calling `super().get_item()` in the wrapper. Since everything is built dynamically, the method "ProtectedStoreAPI::get_movie" is not detected as a `RequestDefinitionBuilder` and is not turned into a callable. Neither is the `StoreAPI::get_movie` since its overriden. Thus, nothing is ever built into a `CallBuilder` and the `StoreAPI` cannot be called.

Note, however, that everything is fine when the protected API calls the method something other than `get_item`, since then `StoreAPI::get_movie` is exposed and can be built.

I've been working with this for the better part of a few hours and can't come up with a good solution. I've tried adding a new `@overrides` decorator and a new Mixin class.. both seemed promising, but no dice.

I think this is a pretty valid usecase and it's a shame it doesn't work. I'll do investigations in my free time and keep this issue updated. Please chime in if you know anything!

### PR DESCRIPTION
**Pull request by brandonio21**
Fixes #23

Changes proposed in this pull request:
- Change CallBuilder to not build upon init()
- RequestDefinitionBuilder builds on access
- Add memozie func

Attention: @prkumar

### CODE REVIEW

File: uplink/utils.py [hunk 1]
```diff
@@ -58,6 +59,19 @@ def get_arg_spec(f):
 except ImportError:
     import urlparse
 
+try:
+    from functools import lru_cache as memoize
+except ImportError:
```
Comment (prkumar) [inline uplink/utils.py]: Looks like the coveralls failure is strictly due to a loss of coverage in this file. Given that it's pretty straightforward memoization, I think it's fair to have `coverage.py` skip this block. Could you add this coverage add  `# pragma: no cover` to this line?

File: uplink/builder.py [hunk 2]
```diff
@@ -160,21 +160,13 @@ def __init__(
             hook=None,
             converter_factories=()
     ):
-        builder = Builder()
-        builder.base_url = base_url
-        builder.add_converter_factory(*converter_factories)
+        self.builder = Builder()
```
Comment (prkumar) [inline uplink/builder.py]: Could you set `self._builder` instead. While python doesn't have a concept of `private` and `public` attributes/methods,  I think it's good practice to use a single leading underscore to omit it from the public API.

File: uplink/utils.py (hunk 1 above)
Comment (brandonio21) [reply uplink/utils.py]: Or why not just add a test? ;)

**Comment by prkumar** [comment]
@brandonio21, Hmm.. I think we still take a coverage hit since the Python 3.3+ tests use `functools.memoize` and never touch the code inside `expect ImportError` block. I wonder if we can tell `coverage.py` to ignore the unused depending on Python version. 

Looks good to me, either way. I'll go ahead an merge this in.

**Comment by brandonio21** [comment]
Thanks for merging. I'm about to propose a change that would make things a bit more testable

**Pull request merged**